local_chainalysis.go

```diff
@@ -0,0 +1,78 @@
+package chainalysis
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/0xsequence/go-sequence/lib/prototyp"
+	"github.com/rs/zerolog/log"
+)
+
+const sanctionedAddressesSource = "https://raw.githubusercontent.com/0xsequence/chainalysis/initial-version/index/sanctioned_addresses.json"
+
+type sanctionedAddressEvent struct {
+	BlockNum  uint64          `json:"blockNum"`
+	BlockHash string          `json:"blockHash"`
+	Addrs     []prototyp.Hash `json:"addrs"`
+}
+
+type localChainalysis struct {
+	mu                  sync.RWMutex
+	SanctionedAddresses map[string]struct{}
+}
+
+func NewLocalChainalysis(ctx context.Context) (Chainalysis, error) {
+	sanctionedAddresses := map[string]struct{}{}
+
+	sanctionedAddressEvents, err := fetchSanctionedAddressEventsFromSource(sanctionedAddressesSource)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, event := range sanctionedAddressEvents {
+		for _, addr := range event.Addrs {
+			sanctionedAddresses[addr.String()] = struct{}{}
+		}
+	}
+
+	lc := &localChainalysis{
+		SanctionedAddresses: sanctionedAddresses,
+	}
+
+	go lc.fetcher(ctx)
+
+	return lc, nil
+}
+
+func (l *localChainalysis) IsSanctioned(address string) (bool, error) {
+	formattedAddress := prototyp.HashFromString(address).String()
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+	_, ok := l.SanctionedAddresses[formattedAddress]
+	return ok, nil
+}
+
+func (l *localChainalysis) fetcher(ctx context.Context) {
+	ticker := time.NewTicker(10 * time.Second)
```

Comment:
lets change this to 5 minutes

the only thing is to ensure it will run on boot, etc.. but afterwards it will only check every 5 or 10 minutes

Reply:
as well... once we fetch the index file... we should also be fetching logs from ethereum to keep synced from the last block we have, using the method `fetchEthereumLogs` / `fetchSanctionedAddressEvents`

the idea is we have a json index which we can fetch on boot very quickly, but if we don't update the file regularly enough, then it will be okay as our runner will get the rest of the addresses from logs off the network. Our index is just like a bootstrap, but the source of truth are the logs on-chain, which is why we cache and push them into our index to bootstrap for faster loading.